GraknClient.java

```diff
@@ -617,9 +625,55 @@ public boolean isClosed() {
             }
         }
 
+        private static class TransactionIterator extends AbstractIterator<SessionProto.Transaction.Iter.Res> implements Transceiver.ResponseReceiver {
+            private final BlockingQueue<Transceiver.Response> queue = new LinkedBlockingQueue<>();
+
+            @Override
+            public boolean onResponse(Transceiver.Response response) {
+                queue.add(response);
+                if (response.type() == OK) {
+                    SessionProto.Transaction.Iter.Res iterRes = response.ok().getIterRes();
+                    switch (iterRes.getResCase()) {
+                        case DONE:
+                        case ITERATORID:
+                            return false;
+                        default:
+                            return true;
+                    }
+                }
+                return false;
+            }
+
+            @Override
+            protected SessionProto.Transaction.Iter.Res computeNext() {
+                Transceiver.Response response;
+                try {
+                    response = queue.take();
+                } catch (InterruptedException e) {
+                    throw GraknConceptException.create("Iteration Interrupted");
+                }
+
+                switch (response.type()) {
+                    case OK:
+                        return response.ok().getIterRes();
+                    case ERROR:
+                        throw GraknClientException.create(response.error().getMessage(), response.error());
+                    case COMPLETED:
+                        throw GraknClientException.create("Transaction interrupted, all running queries have been stopped.");
```

Comment:
Are we sure we're supposed to throw an exception here? Does `COMPLETED` not mean that the stream of responses has "correctly completed"?

Reply:
No, this is when the GRPC stream is completed, which is not expected (should happen after the transaction is closed).

Reply:
right. 👍